tests/functional/test_api.py

```diff
@@ -0,0 +1,181 @@
+import os
+import stat
+import tempfile
+
+from unittest import TestCase
+from requests.exceptions import ConnectionError
+import mock
+from nose.tools import assert_is_none, assert_equal, raises
+
+from miqcli import api
+
+#: token value from tests/assets/auth_token
+AUTH_TOKEN_VALUE = '78asdjasd7nasd90asdmzxc90'
+
+#: fake token file
+TEMP_AUTH_TOKEN = tempfile.NamedTemporaryFile()
+
+#: fake token file without permission to read
+TEMP_AUTH_TOKEN_NOPERMISSION = tempfile.NamedTemporaryFile()
+os.chmod(TEMP_AUTH_TOKEN_NOPERMISSION.name, stat.S_IXUSR)
+
+#: fake value to test setting auth token file
+FAKE_AUTH_TOKEN_VALUE = 'mkzxcasaaxc87z5zxc5zxc5'
+
+
+class TestApi(TestCase):
+    """Test API module"""
+
+    def setUp(self):
+        self.cli_emptysettings = api.ClientAPI({})
+        self.cli_withsettings = api.ClientAPI({
+            'token': 'abcd1234',
+            'username': 'johnny',
+            'password': 'p4ssw0rd',
+            'url': 'https://mydomain.com'
+        })
+        self.temp_auth_file = tempfile.NamedTemporaryFile()
+
+    def tearDown(self):
+        del(self.cli_emptysettings)
+        del(self.cli_withsettings)
+        del(self.temp_auth_file)
+
+    @raises(SystemExit)
+    @mock.patch('miqcli.api.TOKENFILE', '/root/you/can/doit')
+    def test_clientapi_creation_tokenfile_error(self):
+        """Test api.ClientAPI creation when auth token can't be created"""
+        mycli = api.ClientAPI({})
+
+    def test_clientapi_empty_settings(self):
+        """Test api.ClientAPI with empty parameters"""
+        assert_is_none(self.cli_emptysettings._client)
+        assert_equal(self.cli_emptysettings._username, 'admin')
+        assert_equal(self.cli_emptysettings._password, 'smartvm')
+        assert_equal(self.cli_emptysettings._url, 'https://localhost:8443/api')
+        assert_equal(self.cli_emptysettings._verify_ssl, False)
+        assert_equal(self.cli_emptysettings._token, None)
+
+    def test_clientapi_client_property(self):
+        """Test api.ClientAPI.client property"""
+        assert_equal(
+            self.cli_emptysettings._client, self.cli_emptysettings.client)
+
+    def test_clientapi_empty_settings_token_property(self):
+        """Test api.ClientAPI.token property without parameters"""
+        assert_is_none(self.cli_emptysettings._client)
+        assert_equal(self.cli_emptysettings._token, None)
+        assert_equal(self.cli_emptysettings.token, None)
+
+    def test_clientapi_token_property(self):
+        """Test api.ClientAPI.token property without parameters"""
+        assert_is_none(self.cli_emptysettings._client)
+        assert_equal(self.cli_emptysettings._token, None)
+        assert_equal(self.cli_emptysettings.token, None)
```

Comment:
These two tests (test_clientapi_empty_settings_token_property and test_clientapi_token_property) are identical.

Reply:
hmmm, yeah, strange why I've done that. I think I wanted to do what I achieve with the test below. I'm removing it. Thanks Vimal.